File: src/main/scala/com/yogpc/qp/render/RenderAdvQuarry.scala
package com.yogpc.qp.render

import com.yogpc.qp.machines.advquarry.AdvQuarryWork.MakeFrame
import com.yogpc.qp.machines.advquarry.{AdvQuarryWork, TileAdvQuarry}
import com.yogpc.qp.machines.base.Area
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.BufferBuilder
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraftforge.client.model.animation.TileEntityRendererFast

object RenderAdvQuarry extends TileEntityRendererFast[TileAdvQuarry] {
  private[this] final val d = 1d / 16d
  val instance: RenderAdvQuarry.type = this
  lazy val sprite: TextureAtlasSprite = Sprites.getMap(Symbol("yellow"))

  override def renderTileEntityFast(te: TileAdvQuarry, x: Double, y: Double, z: Double,
                                    partialTicks: Float, destroyStage: Int, buffer: BufferBuilder): Unit = {

    Minecraft.getInstance.getProfiler.startSection("quarryplus")

    if (te.action.isInstanceOf[MakeFrame] || (te.action == AdvQuarryWork.waiting)) {
      Minecraft.getInstance.getProfiler.startSection("chunkdestroyer")
      val range = te.area
      if (range != Area.zeroArea) {
        val pos = te.getPos
        val playerX = pos.getX - x
        val playerY = pos.getY - y
        val playerZ = pos.getZ - z
        buffer.setTranslation(-playerX, -playerY, -playerZ)
        val b1 = (playerZ - range.zMin - 0.5).abs < 256
        val b2 = (playerZ - range.zMax + 1.5).abs < 256
        val b3 = (playerX - range.xMin - 0.5).abs < 256
        val b4 = (playerX - range.xMax + 1.5).abs < 256
        val xMin = math.max(range.xMin - 0.5, playerX - 128)
        val xMax = math.min(range.xMax + 1.5, playerX + 128)
        val zMin = math.max(range.zMin - 0.5, playerZ - 128)
        val zMax = math.min(range.zMax + 1.5, playerZ + 128)
        if (b1) Box(xMin, range.yMin, range.zMin - 0.5, xMax, range.yMax, range.zMin - 0.5, d, d, d, firstSide = false, endSide = false).render(buffer, sprite)
        if (b2) Box(xMin, range.yMin, range.zMax + 1.5, xMax, range.yMax, range.zMax + 1.5, d, d, d, firstSide = false, endSide = false).render(buffer, sprite)
        if (b3) Box(range.xMin - 0.5, range.yMin, zMin, range.xMin - 0.5, range.yMax, zMax, d, d, d, firstSide = false, endSide = false).render(buffer, sprite)
        if (b4) Box(range.xMax + 1.5, range.yMin, zMin, range.xMax + 1.5, range.yMax, zMax, d, d, d, firstSide = false, endSide = false).render(buffer, sprite)
        buffer.setTranslation(0, 0, 0)
      }
      Minecraft.getInstance.getProfiler.endSection()
    }

    Minecraft.getInstance.getProfiler.endSection()
  }

  override def isGlobalRenderer(te: TileAdvQuarry): Boolean = true

}

File: src/main/scala/com/yogpc/qp/render/RenderAdvQuarry.java
package com.yogpc.qp.render;

import com.mojang.blaze3d.vertex.PoseStack;
import com.yogpc.qp.QuarryPlus;
import com.yogpc.qp.machines.advquarry.BlockAdvQuarry;
import com.yogpc.qp.machines.advquarry.TileAdvQuarry;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;

@Environment(EnvType.CLIENT)
public class RenderAdvQuarry implements BlockEntityRenderer<TileAdvQuarry> {
    @SuppressWarnings("unused")
    public RenderAdvQuarry(BlockEntityRendererProvider.Context context) {
    }

    @Override
    public boolean shouldRenderOffScreen(TileAdvQuarry blockEntity) {
        return true;
    }

    @Override
    public int getViewDistance() {
        return 256;
    }

    @Override
    @SuppressWarnings("DuplicatedCode") // for readability.
    public void render(TileAdvQuarry quarry, float tickDelta, PoseStack matrices, MultiBufferSource vertexConsumers, int light, int overlay) {
        Minecraft.getInstance().getProfiler().push(QuarryPlus.modID);

        if ("MakeFrame".equals(quarry.actionKey) || "Waiting".equals(quarry.actionKey)) {
            Minecraft.getInstance().getProfiler().push(BlockAdvQuarry.NAME);
            var range = quarry.getArea();
            if (range != null) {
                final double d = 1d / 16d;
                final TextureAtlasSprite sprite = Sprites.INSTANCE.getWhite();
                final ColorBox color = new ColorBox(0xFF, 0xFF, 0, 0xFF);
                var buffer = vertexConsumers.getBuffer(RenderType.cutout());
                var pos = quarry.getBlockPos();
                var player = Minecraft.getInstance().player;
                var playerX = player == null ? pos.getX() : player.getX(); //x
                var playerZ = player == null ? pos.getZ() : player.getZ(); //z
                matrices.pushPose();
                matrices.translate(-pos.getX(), -pos.getY(), -pos.getZ()); // Offset
                var startX = range.minX() + 0.5;
                var startZ = range.minZ() + 0.5;
                var endZ = range.maxZ() + 0.5;
                var endX = range.maxX() + 0.5;
                var b1 = Math.abs(playerZ - startZ) < 256;
                var b2 = Math.abs(playerZ - endZ) < 256;
                var b3 = Math.abs(playerX - startX) < 256;
                var b4 = Math.abs(playerX - endX) < 256;
                var xMin = Math.max(startX, playerX - 128);
                var xMax = Math.min(endX, playerX + 128);
                var zMin = Math.max(startZ, playerZ - 128);
                var zMax = Math.min(endZ, playerZ + 128);
                if (b1)
                    Box.apply(xMin, range.minY(), startZ, xMax, range.minY(), startZ, xMax - xMin, d, d, false, false).render(buffer, matrices, sprite, color);
                if (b2)
                    Box.apply(xMin, range.minY(), endZ, xMax, range.minY(), endZ, xMax - xMin, d, d, false, false).render(buffer, matrices, sprite, color);
                if (b3)
                    Box.apply(startX, range.minY(), zMin, startX, range.minY(), zMax, d, d, zMax - zMin, false, false).render(buffer, matrices, sprite, color);
                if (b4)
                    Box.apply(endX, range.minY(), zMin, endX, range.minY(), zMax, d, d, zMax - zMin, false, false).render(buffer, matrices, sprite, color);
                matrices.popPose();
            }
            Minecraft.getInstance().getProfiler().pop();
        }

        Minecraft.getInstance().getProfiler().pop();
    }
}
